src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java

```diff
@@ -0,0 +1,390 @@
+    {
+        @Nullable final AbstractBuilding building = getBuilding();
+
+        boolean hasItem;
+        if(building != null)
+        {
+            if(tool.equals(Utils.PICKAXE))
+            {
+                hasItem = InventoryUtils.isPickaxeInTileEntity(building.getTileEntity(), building.getNeededPickaxeLevel());
+            }
+            else
+            {
+                hasItem = InventoryUtils.isToolInTileEntity(building.getTileEntity(), tool);
+            }
+
+            if(hasItem)
+            {
+                return true;
+            }
+
+            for(BlockPos pos : building.getAdditionalCountainers())
+            {
+                TileEntity entity = worldObj.getTileEntity(pos);
+                if(entity instanceof TileEntityChest)
+                {
+                    if(tool.equals(Utils.PICKAXE))
+                    {
+                        hasItem = InventoryUtils.isPickaxeInTileEntity(building.getTileEntity(), building.getNeededPickaxeLevel());
+                    }
+                    else
+                    {
+                        hasItem = InventoryUtils.isToolInTileEntity((TileEntityChest) entity, tool);
+                    }
+
+                    if(hasItem)
+                    {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+
+    /**
+     * Finds the first @see ItemStack the type of {@code is}.
+     * It will be taken from the chest and placed in the workers inventory.
+     * Make sure that the worker stands next the chest to not break immersion.
+     * Also make sure to have inventory space for the stack.
+     * @param entity the tileEntity chest or building.
+     * @param is the itemStack.
+     * @return true if found the stack.
+     */
+    public boolean isInTileEntity(TileEntityChest entity, ItemStack is)
+    {
+        return is != null
+                && InventoryFunctions
+                .matchFirstInInventoryWithInventory(
+                        entity,
+                        stack -> stack != null && is.isItemEqual(stack),
+                        InventoryFunctions::doNothing
+                );
+    }
+
+    @Override
+    public void readFromNBT(final NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+    }
+
+    @NotNull
+    @Override
+    public NBTTagCompound writeToNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.writeToNBT(compound);
+        return compound;
+    }
+
+    /**
+     * Dump the inventory of a citizen into the warehouse.
+     * Go through all items and search the right chest to dump it in.
+     * @param inventoryCitizen
+     */
+    public void dumpInventoryIntoWareHouse(@NotNull final InventoryCitizen inventoryCitizen)
+    {
+        for(int i = 0; i < inventoryCitizen.getSizeInventory(); i++)
+        {
+            ItemStack stack = inventoryCitizen.getStackInSlot(i);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'stack' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)